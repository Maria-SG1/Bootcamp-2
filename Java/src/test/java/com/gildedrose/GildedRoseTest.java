package com.gildedrose;

import static org.junit.jupiter.api.Assertions.assertEquals;
import java.util.List;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource; 
    public class GildedRoseTest {

		private Item[] createItem(String name, int sellIn, int quality) {
	        return new Item[]{ new Item(name, sellIn, quality) };
	    }
	    private void updateAndAssert(Item[] items, String expected) {
	        GildedRose app = new GildedRose(items);
	        app.updateQuality();
	        assertEquals(expected, app.items[0].toString());
	    }

	    @Nested
	    @DisplayName("Elixir of the Mongoose")
	    class Elixir {
	        @ParameterizedTest
	        @CsvSource({
	            "10, 30, Elixir of the Mongoose, 9, 29",
	            "10, 0, Elixir of the Mongoose, 9, 0"
	        })
	        void testElixir(int sellIn, int quality, String expectedName, int expectedSellIn, int expectedQuality) {
	            updateAndAssert(createItem("Elixir of the Mongoose", sellIn, quality), 
	                            expectedName + ", " + expectedSellIn + ", " + expectedQuality);
	        }
	    }

	    @Nested
	    @DisplayName("Aged Brie")
	    class AgedBrie {
	        @ParameterizedTest
	        @CsvSource({
	            "10, 50, Aged Brie, 9, 50",
	            "2, 0, Aged Brie, 1, 1",
	            "-1, 40, Aged Brie, -2, 42",
	            "-1, 50, Aged Brie, -2, 50"
	        })
	        void testAgedBrie(int sellIn, int quality, String expectedName, int expectedSellIn, int expectedQuality) {
	            updateAndAssert(createItem("Aged Brie", sellIn, quality), 
	                            expectedName + ", " + expectedSellIn + ", " + expectedQuality);
	        }
	    }

	    @Nested
	    @DisplayName("Backstage passes")
	    class Backstage {
	        @ParameterizedTest
	        @CsvSource({
	            "9, 40, Backstage passes to a TAFKAL80ETC concert, 8, 42",
	            "11, 40, Backstage passes to a TAFKAL80ETC concert, 10, 41",
	            "10, 50, Backstage passes to a TAFKAL80ETC concert, 9, 50",
	            "4, 40, Backstage passes to a TAFKAL80ETC concert, 3, 43",
	            "0, 40, Backstage passes to a TAFKAL80ETC concert, -1, 0"
	        })
	        void testBackstage(int sellIn, int quality, String expectedName, int expectedSellIn, int expectedQuality) {
	            updateAndAssert(createItem("Backstage passes to a TAFKAL80ETC concert", sellIn, quality), 
	                            expectedName + ", " + expectedSellIn + ", " + expectedQuality);
	        }
	    }

	    @Nested
	    @DisplayName("Otros productos")
	    class Otros {
	        @ParameterizedTest
	        @CsvSource({
	            "-1, 10, +5 Dexterity Vest, -2, 8",
	            "-1, 80, 'Sulfuras, Hand of Ragnaros', -1, 80",
	            "5, 80, 'Sulfuras, Hand of Ragnaros', 5, 80",
	            "5, 60, 'Sulfuras, Hand of Ragnaros', 5, 60"
	        })
	        void testOtros(int sellIn, int quality, String expectedName, int expectedSellIn, int expectedQuality) {
	            updateAndAssert(createItem(expectedName, sellIn, quality), 
	                            expectedName + ", " + expectedSellIn + ", " + expectedQuality);
	        }
	    }
	    
	    @Nested
	    @DisplayName("Sulfuras, Hand of Ragnaros")
	    class Sulfuras {
	        @ParameterizedTest
	        @CsvSource({
	            "-1, 80, 'Sulfuras, Hand of Ragnaros', -1, 80",
	            "5, 80, 'Sulfuras, Hand of Ragnaros', 5, 80",
	            "5, 60, 'Sulfuras, Hand of Ragnaros', 5, 60"	            
	        })
	        void testSulfuras(int sellIn, int quality, String expectedName, int expectedSellIn, int expectedQuality) {
	            updateAndAssert(createItem(expectedName, sellIn, quality), 
	                            expectedName + ", " + expectedSellIn + ", " + expectedQuality);
	        }
	    }

	    @Nested
	    @DisplayName("Conjured")
	    class Conjured {
	        @ParameterizedTest
	        @CsvSource({
	            "3, 10, Conjured Mana Cake, 2, 8",
	            "-1, 10, Conjured Mana Cake, -2, 6",
	            "3, 1, Conjured Mana Cake, 2, 0"
	        })
	        void testConjured(int sellIn, int quality, String expectedName, int expectedSellIn, int expectedQuality) {
	            updateAndAssert(createItem("Conjured Mana Cake", sellIn, quality), 
	                            expectedName + ", " + expectedSellIn + ", " + expectedQuality);
	        }
	    }
	    
	    /* Utilizando Approvals.verify */
	    
	    @Test
//		@Disabled
		void instantanea() {
			Item[] items = new Item[] { 
					new Item("+5 Dexterity Vest", 10, 20), 
					new Item("Aged Brie", 2, 0),
					new Item("Elixir of the Mongoose", 5, 7), 
					new Item("Sulfuras, Hand of Ragnaros", 0, 80),
					new Item("Sulfuras, Hand of Ragnaros", -1, 80),
					new Item("Backstage passes to a TAFKAL80ETC concert", 15, 20),
					new Item("Backstage passes to a TAFKAL80ETC concert", 10, 49),
					new Item("Backstage passes to a TAFKAL80ETC concert", 5, 49),
					// this conjured item does not work properly yet 
					new Item("Conjured Mana Cake", 3, 6) 
					};

			GildedRose app = new GildedRose(items);
			StringBuilder output = new StringBuilder();
			output.append("day,name, sellIn, quality\n");
			List.of(items).forEach(item -> output.append("0," + item + "\n"));
			for (int i = 1; i <= 31; i++) {
				app.updateQuality();
				for(Item item: items)
					output.append(i + "," + item + "\n");
			}
			Approvals.verify(output);
	 }   
}


