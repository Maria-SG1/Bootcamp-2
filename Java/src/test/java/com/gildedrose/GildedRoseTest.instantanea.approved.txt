day,name, sellIn, quality
0,+5 Dexterity Vest, 10, 20
0,Aged Brie, 2, 0
0,Elixir of the Mongoose, 5, 7
0,Sulfuras, Hand of Ragnaros, 0, 80
0,Sulfuras, Hand of Ragnaros, -1, 80
0,Backstage passes to a TAFKAL80ETC concert, 15, 20
0,Backstage passes to a TAFKAL80ETC concert, 10, 49
0,Backstage passes to a TAFKAL80ETC concert, 5, 49
0,Conjured Mana Cake, 3, 6
1,+5 Dexterity Vest, 9, 19
1,Aged Brie, 2, 0
1,Elixir of the Mongoose, 5, 7
1,Sulfuras, Hand of Ragnaros, 0, 80
1,Sulfuras, Hand of Ragnaros, -1, 80
1,Backstage passes to a TAFKAL80ETC concert, 15, 20
1,Backstage passes to a TAFKAL80ETC concert, 10, 49
1,Backstage passes to a TAFKAL80ETC concert, 5, 49
1,Conjured Mana Cake, 3, 6
2,+5 Dexterity Vest, 8, 18
2,Aged Brie, 2, 0
2,Elixir of the Mongoose, 5, 7
2,Sulfuras, Hand of Ragnaros, 0, 80
2,Sulfuras, Hand of Ragnaros, -1, 80
2,Backstage passes to a TAFKAL80ETC concert, 15, 20
2,Backstage passes to a TAFKAL80ETC concert, 10, 49
2,Backstage passes to a TAFKAL80ETC concert, 5, 49
2,Conjured Mana Cake, 3, 6
3,+5 Dexterity Vest, 7, 17
3,Aged Brie, 2, 0
3,Elixir of the Mongoose, 5, 7
3,Sulfuras, Hand of Ragnaros, 0, 80
3,Sulfuras, Hand of Ragnaros, -1, 80
3,Backstage passes to a TAFKAL80ETC concert, 15, 20
3,Backstage passes to a TAFKAL80ETC concert, 10, 49
3,Backstage passes to a TAFKAL80ETC concert, 5, 49
3,Conjured Mana Cake, 3, 6
4,+5 Dexterity Vest, 6, 16
4,Aged Brie, 2, 0
4,Elixir of the Mongoose, 5, 7
4,Sulfuras, Hand of Ragnaros, 0, 80
4,Sulfuras, Hand of Ragnaros, -1, 80
4,Backstage passes to a TAFKAL80ETC concert, 15, 20
4,Backstage passes to a TAFKAL80ETC concert, 10, 49
4,Backstage passes to a TAFKAL80ETC concert, 5, 49
4,Conjured Mana Cake, 3, 6
5,+5 Dexterity Vest, 5, 15
5,Aged Brie, 2, 0
5,Elixir of the Mongoose, 5, 7
5,Sulfuras, Hand of Ragnaros, 0, 80
5,Sulfuras, Hand of Ragnaros, -1, 80
5,Backstage passes to a TAFKAL80ETC concert, 15, 20
5,Backstage passes to a TAFKAL80ETC concert, 10, 49
5,Backstage passes to a TAFKAL80ETC concert, 5, 49
5,Conjured Mana Cake, 3, 6
6,+5 Dexterity Vest, 4, 14
6,Aged Brie, 2, 0
6,Elixir of the Mongoose, 5, 7
6,Sulfuras, Hand of Ragnaros, 0, 80
6,Sulfuras, Hand of Ragnaros, -1, 80
6,Backstage passes to a TAFKAL80ETC concert, 15, 20
6,Backstage passes to a TAFKAL80ETC concert, 10, 49
6,Backstage passes to a TAFKAL80ETC concert, 5, 49
6,Conjured Mana Cake, 3, 6
7,+5 Dexterity Vest, 3, 13
7,Aged Brie, 2, 0
7,Elixir of the Mongoose, 5, 7
7,Sulfuras, Hand of Ragnaros, 0, 80
7,Sulfuras, Hand of Ragnaros, -1, 80
7,Backstage passes to a TAFKAL80ETC concert, 15, 20
7,Backstage passes to a TAFKAL80ETC concert, 10, 49
7,Backstage passes to a TAFKAL80ETC concert, 5, 49
7,Conjured Mana Cake, 3, 6
8,+5 Dexterity Vest, 2, 12
8,Aged Brie, 2, 0
8,Elixir of the Mongoose, 5, 7
8,Sulfuras, Hand of Ragnaros, 0, 80
8,Sulfuras, Hand of Ragnaros, -1, 80
8,Backstage passes to a TAFKAL80ETC concert, 15, 20
8,Backstage passes to a TAFKAL80ETC concert, 10, 49
8,Backstage passes to a TAFKAL80ETC concert, 5, 49
8,Conjured Mana Cake, 3, 6
9,+5 Dexterity Vest, 1, 11
9,Aged Brie, 2, 0
9,Elixir of the Mongoose, 5, 7
9,Sulfuras, Hand of Ragnaros, 0, 80
9,Sulfuras, Hand of Ragnaros, -1, 80
9,Backstage passes to a TAFKAL80ETC concert, 15, 20
9,Backstage passes to a TAFKAL80ETC concert, 10, 49
9,Backstage passes to a TAFKAL80ETC concert, 5, 49
9,Conjured Mana Cake, 3, 6
10,+5 Dexterity Vest, 0, 10
10,Aged Brie, 2, 0
10,Elixir of the Mongoose, 5, 7
10,Sulfuras, Hand of Ragnaros, 0, 80
10,Sulfuras, Hand of Ragnaros, -1, 80
10,Backstage passes to a TAFKAL80ETC concert, 15, 20
10,Backstage passes to a TAFKAL80ETC concert, 10, 49
10,Backstage passes to a TAFKAL80ETC concert, 5, 49
10,Conjured Mana Cake, 3, 6
11,+5 Dexterity Vest, -1, 9
11,Aged Brie, 2, 0
11,Elixir of the Mongoose, 5, 7
11,Sulfuras, Hand of Ragnaros, 0, 80
11,Sulfuras, Hand of Ragnaros, -1, 80
11,Backstage passes to a TAFKAL80ETC concert, 15, 20
11,Backstage passes to a TAFKAL80ETC concert, 10, 49
11,Backstage passes to a TAFKAL80ETC concert, 5, 49
11,Conjured Mana Cake, 3, 6
12,+5 Dexterity Vest, -2, 7
12,Aged Brie, 2, 0
12,Elixir of the Mongoose, 5, 7
12,Sulfuras, Hand of Ragnaros, 0, 80
12,Sulfuras, Hand of Ragnaros, -1, 80
12,Backstage passes to a TAFKAL80ETC concert, 15, 20
12,Backstage passes to a TAFKAL80ETC concert, 10, 49
12,Backstage passes to a TAFKAL80ETC concert, 5, 49
12,Conjured Mana Cake, 3, 6
13,+5 Dexterity Vest, -3, 5
13,Aged Brie, 2, 0
13,Elixir of the Mongoose, 5, 7
13,Sulfuras, Hand of Ragnaros, 0, 80
13,Sulfuras, Hand of Ragnaros, -1, 80
13,Backstage passes to a TAFKAL80ETC concert, 15, 20
13,Backstage passes to a TAFKAL80ETC concert, 10, 49
13,Backstage passes to a TAFKAL80ETC concert, 5, 49
13,Conjured Mana Cake, 3, 6
14,+5 Dexterity Vest, -4, 3
14,Aged Brie, 2, 0
14,Elixir of the Mongoose, 5, 7
14,Sulfuras, Hand of Ragnaros, 0, 80
14,Sulfuras, Hand of Ragnaros, -1, 80
14,Backstage passes to a TAFKAL80ETC concert, 15, 20
14,Backstage passes to a TAFKAL80ETC concert, 10, 49
14,Backstage passes to a TAFKAL80ETC concert, 5, 49
14,Conjured Mana Cake, 3, 6
15,+5 Dexterity Vest, -5, 1
15,Aged Brie, 2, 0
15,Elixir of the Mongoose, 5, 7
15,Sulfuras, Hand of Ragnaros, 0, 80
15,Sulfuras, Hand of Ragnaros, -1, 80
15,Backstage passes to a TAFKAL80ETC concert, 15, 20
15,Backstage passes to a TAFKAL80ETC concert, 10, 49
15,Backstage passes to a TAFKAL80ETC concert, 5, 49
15,Conjured Mana Cake, 3, 6
16,+5 Dexterity Vest, -6, -1
16,Aged Brie, 2, 0
16,Elixir of the Mongoose, 5, 7
16,Sulfuras, Hand of Ragnaros, 0, 80
16,Sulfuras, Hand of Ragnaros, -1, 80
16,Backstage passes to a TAFKAL80ETC concert, 15, 20
16,Backstage passes to a TAFKAL80ETC concert, 10, 49
16,Backstage passes to a TAFKAL80ETC concert, 5, 49
16,Conjured Mana Cake, 3, 6
17,+5 Dexterity Vest, -7, 0
17,Aged Brie, 2, 0
17,Elixir of the Mongoose, 5, 7
17,Sulfuras, Hand of Ragnaros, 0, 80
17,Sulfuras, Hand of Ragnaros, -1, 80
17,Backstage passes to a TAFKAL80ETC concert, 15, 20
17,Backstage passes to a TAFKAL80ETC concert, 10, 49
17,Backstage passes to a TAFKAL80ETC concert, 5, 49
17,Conjured Mana Cake, 3, 6
18,+5 Dexterity Vest, -8, 0
18,Aged Brie, 2, 0
18,Elixir of the Mongoose, 5, 7
18,Sulfuras, Hand of Ragnaros, 0, 80
18,Sulfuras, Hand of Ragnaros, -1, 80
18,Backstage passes to a TAFKAL80ETC concert, 15, 20
18,Backstage passes to a TAFKAL80ETC concert, 10, 49
18,Backstage passes to a TAFKAL80ETC concert, 5, 49
18,Conjured Mana Cake, 3, 6
19,+5 Dexterity Vest, -9, 0
19,Aged Brie, 2, 0
19,Elixir of the Mongoose, 5, 7
19,Sulfuras, Hand of Ragnaros, 0, 80
19,Sulfuras, Hand of Ragnaros, -1, 80
19,Backstage passes to a TAFKAL80ETC concert, 15, 20
19,Backstage passes to a TAFKAL80ETC concert, 10, 49
19,Backstage passes to a TAFKAL80ETC concert, 5, 49
19,Conjured Mana Cake, 3, 6
20,+5 Dexterity Vest, -10, 0
20,Aged Brie, 2, 0
20,Elixir of the Mongoose, 5, 7
20,Sulfuras, Hand of Ragnaros, 0, 80
20,Sulfuras, Hand of Ragnaros, -1, 80
20,Backstage passes to a TAFKAL80ETC concert, 15, 20
20,Backstage passes to a TAFKAL80ETC concert, 10, 49
20,Backstage passes to a TAFKAL80ETC concert, 5, 49
20,Conjured Mana Cake, 3, 6
21,+5 Dexterity Vest, -11, 0
21,Aged Brie, 2, 0
21,Elixir of the Mongoose, 5, 7
21,Sulfuras, Hand of Ragnaros, 0, 80
21,Sulfuras, Hand of Ragnaros, -1, 80
21,Backstage passes to a TAFKAL80ETC concert, 15, 20
21,Backstage passes to a TAFKAL80ETC concert, 10, 49
21,Backstage passes to a TAFKAL80ETC concert, 5, 49
21,Conjured Mana Cake, 3, 6
22,+5 Dexterity Vest, -12, 0
22,Aged Brie, 2, 0
22,Elixir of the Mongoose, 5, 7
22,Sulfuras, Hand of Ragnaros, 0, 80
22,Sulfuras, Hand of Ragnaros, -1, 80
22,Backstage passes to a TAFKAL80ETC concert, 15, 20
22,Backstage passes to a TAFKAL80ETC concert, 10, 49
22,Backstage passes to a TAFKAL80ETC concert, 5, 49
22,Conjured Mana Cake, 3, 6
23,+5 Dexterity Vest, -13, 0
23,Aged Brie, 2, 0
23,Elixir of the Mongoose, 5, 7
23,Sulfuras, Hand of Ragnaros, 0, 80
23,Sulfuras, Hand of Ragnaros, -1, 80
23,Backstage passes to a TAFKAL80ETC concert, 15, 20
23,Backstage passes to a TAFKAL80ETC concert, 10, 49
23,Backstage passes to a TAFKAL80ETC concert, 5, 49
23,Conjured Mana Cake, 3, 6
24,+5 Dexterity Vest, -14, 0
24,Aged Brie, 2, 0
24,Elixir of the Mongoose, 5, 7
24,Sulfuras, Hand of Ragnaros, 0, 80
24,Sulfuras, Hand of Ragnaros, -1, 80
24,Backstage passes to a TAFKAL80ETC concert, 15, 20
24,Backstage passes to a TAFKAL80ETC concert, 10, 49
24,Backstage passes to a TAFKAL80ETC concert, 5, 49
24,Conjured Mana Cake, 3, 6
25,+5 Dexterity Vest, -15, 0
25,Aged Brie, 2, 0
25,Elixir of the Mongoose, 5, 7
25,Sulfuras, Hand of Ragnaros, 0, 80
25,Sulfuras, Hand of Ragnaros, -1, 80
25,Backstage passes to a TAFKAL80ETC concert, 15, 20
25,Backstage passes to a TAFKAL80ETC concert, 10, 49
25,Backstage passes to a TAFKAL80ETC concert, 5, 49
25,Conjured Mana Cake, 3, 6
26,+5 Dexterity Vest, -16, 0
26,Aged Brie, 2, 0
26,Elixir of the Mongoose, 5, 7
26,Sulfuras, Hand of Ragnaros, 0, 80
26,Sulfuras, Hand of Ragnaros, -1, 80
26,Backstage passes to a TAFKAL80ETC concert, 15, 20
26,Backstage passes to a TAFKAL80ETC concert, 10, 49
26,Backstage passes to a TAFKAL80ETC concert, 5, 49
26,Conjured Mana Cake, 3, 6
27,+5 Dexterity Vest, -17, 0
27,Aged Brie, 2, 0
27,Elixir of the Mongoose, 5, 7
27,Sulfuras, Hand of Ragnaros, 0, 80
27,Sulfuras, Hand of Ragnaros, -1, 80
27,Backstage passes to a TAFKAL80ETC concert, 15, 20
27,Backstage passes to a TAFKAL80ETC concert, 10, 49
27,Backstage passes to a TAFKAL80ETC concert, 5, 49
27,Conjured Mana Cake, 3, 6
28,+5 Dexterity Vest, -18, 0
28,Aged Brie, 2, 0
28,Elixir of the Mongoose, 5, 7
28,Sulfuras, Hand of Ragnaros, 0, 80
28,Sulfuras, Hand of Ragnaros, -1, 80
28,Backstage passes to a TAFKAL80ETC concert, 15, 20
28,Backstage passes to a TAFKAL80ETC concert, 10, 49
28,Backstage passes to a TAFKAL80ETC concert, 5, 49
28,Conjured Mana Cake, 3, 6
29,+5 Dexterity Vest, -19, 0
29,Aged Brie, 2, 0
29,Elixir of the Mongoose, 5, 7
29,Sulfuras, Hand of Ragnaros, 0, 80
29,Sulfuras, Hand of Ragnaros, -1, 80
29,Backstage passes to a TAFKAL80ETC concert, 15, 20
29,Backstage passes to a TAFKAL80ETC concert, 10, 49
29,Backstage passes to a TAFKAL80ETC concert, 5, 49
29,Conjured Mana Cake, 3, 6
30,+5 Dexterity Vest, -20, 0
30,Aged Brie, 2, 0
30,Elixir of the Mongoose, 5, 7
30,Sulfuras, Hand of Ragnaros, 0, 80
30,Sulfuras, Hand of Ragnaros, -1, 80
30,Backstage passes to a TAFKAL80ETC concert, 15, 20
30,Backstage passes to a TAFKAL80ETC concert, 10, 49
30,Backstage passes to a TAFKAL80ETC concert, 5, 49
30,Conjured Mana Cake, 3, 6
31,+5 Dexterity Vest, -21, 0
31,Aged Brie, 2, 0
31,Elixir of the Mongoose, 5, 7
31,Sulfuras, Hand of Ragnaros, 0, 80
31,Sulfuras, Hand of Ragnaros, -1, 80
31,Backstage passes to a TAFKAL80ETC concert, 15, 20
31,Backstage passes to a TAFKAL80ETC concert, 10, 49
31,Backstage passes to a TAFKAL80ETC concert, 5, 49
31,Conjured Mana Cake, 3, 6
